cpp/src/arrow/array.cc

```diff
@@ -482,6 +482,34 @@ StructArray::StructArray(const std::shared_ptr<DataType>& type, int64_t length,
   boxed_fields_.resize(children.size());
 }
 
+Result<std::shared_ptr<Array>> StructArray::Make(
+    const std::vector<std::shared_ptr<Array>>& children,
+    const std::vector<std::string>& field_names, std::shared_ptr<Buffer> null_bitmap,
```

Comment:
I think it would be better to make the association between field names and child arrays more clear in the arguments:

```suggestion
    const std::vector<std::pair<std::string, std::shared_ptr<Array>>>& children,
    std::shared_ptr<Buffer> null_bitmap,
```

This would allow usage like:
```c++
auto s = StructArray::Make({{"a", a}, {"b", b}});
```

Reply:
As a side effect, this would make a having more field names than child arrays a compile-time error

Reply:
We're already using this idiom elsewhere though (separate arrays and names vectors).

Reply:
OK, we can always add another helper if we want :)